File: src/main/java/org/magicghostvu/actorvt/context/ActorContext.java
package org.magicghostvu.actorvt.context;

import org.magicghostvu.actorvt.ActorRef;
import org.magicghostvu.actorvt.behavior.Behavior;

import java.util.concurrent.ConcurrentHashMap;

public sealed abstract class ActorContext permits ActorSystem, GeneralActorContext {

    public abstract String getPath();

    final ConcurrentHashMap<ActorRef<?>, GeneralActorContext<?>> refToChild = new ConcurrentHashMap<>();

    public abstract <Protocol> ActorRef<Protocol> spawn(
            String childName,
            int queueCapacity,
            ActorVTSupplier<Behavior<Protocol>> behaviorFactory
    );

    public abstract void stopChild(ActorRef<?> actorRef);
}

File: src/main/kotlin/org/magicghostvu/actorvt/context/ActorContext.kt
package org.magicghostvu.actorvt.context

import org.magicghostvu.actorvt.ActorRef
import org.magicghostvu.actorvt.behavior.Behavior
import java.util.concurrent.ConcurrentHashMap

sealed class ActorContext {


    internal abstract val path: String;

    // name -> child
    internal val refToChild: MutableMap<ActorRef<*>, GeneralActorContext<*>> = ConcurrentHashMap()


    // expect func này chỉ được gọi trong actor thread
    // todo: re-design this method signature
    abstract fun <Protocol> spawn(
        childName: String,
        queueCapacity: Int,
        behaviorFactory: () -> Behavior<Protocol>
    ): ActorRef<Protocol>;

    abstract fun stopChild(actorRef: ActorRef<*>);
}